// !WITH_NEW_INFERENCE
package a

class MyClass {
    fun component1(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>i<!><!>: Int) {}
}

class MyClass2 {}

<!NI;CONFLICTING_OVERLOADS!><!CONFLICTING_OVERLOADS!>fun MyClass2.component1()<!><!> = 1.2
<!NI;CONFLICTING_OVERLOADS!><!CONFLICTING_OVERLOADS!>fun MyClass2.component1()<!><!> = 1.3

fun test(mc1: MyClass, mc2: MyClass2) {
    val (<!OPERATOR_MODIFIER_REQUIRED!>a<!>, b) = <!NI;COMPONENT_FUNCTION_MISSING!><!NI;COMPONENT_FUNCTION_MISSING!><!COMPONENT_FUNCTION_MISSING!><!COMPONENT_FUNCTION_MISSING!>mc1<!><!><!><!>
    val (c) = <!NI;COMPONENT_FUNCTION_AMBIGUITY!><!COMPONENT_FUNCTION_AMBIGUITY!>mc2<!><!>

    //a,b,c are error types
    use(<!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>a<!><!>, <!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>b<!><!>, <!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>c<!><!>)
}

fun use(vararg a: Any?) = a