// !WITH_NEW_INFERENCE
fun foo(): Int {
    var i: Int? = <!NI;VARIABLE_WITH_REDUNDANT_INITIALIZER!><!VARIABLE_WITH_REDUNDANT_INITIALIZER!>42<!><!>
    i = null
    return <!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!><!DEBUG_INFO_CONSTANT!>i<!> + 1<!><!>
}