// !WITH_NEW_INFERENCE
fun test(): Array<Int> {
    <!NI;UNSUPPORTED!><!UNSUPPORTED!>[1, 2]<!><!>
    <!NI;UNSUPPORTED!><!UNSUPPORTED!>[1, 2]<!><!>[0]
    <!NI;UNSUPPORTED!><!UNSUPPORTED!>[1, 2]<!><!>.get(0)

    foo(<!NI;UNSUPPORTED!><!UNSUPPORTED!>[""]<!><!>)

    val <!NI;UNUSED_VARIABLE!><!UNUSED_VARIABLE!>p<!><!> = <!NI;UNSUPPORTED!><!UNSUPPORTED!>[1, 2]<!><!> <!NI;UNRESOLVED_REFERENCE_WRONG_RECEIVER!><!NI;DEBUG_INFO_UNRESOLVED_WITH_TARGET!><!UNRESOLVED_REFERENCE_WRONG_RECEIVER!>+<!><!><!> <!UNSUPPORTED!>[3, 4]<!>

    return <!NI;UNSUPPORTED!><!UNSUPPORTED!>[1, 2]<!><!>
}

fun foo(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>a<!><!>: Array<String> = <!NI;UNSUPPORTED!><!UNSUPPORTED!>[""]<!><!>) {}

class A(val a: Array<Int> = <!NI;UNSUPPORTED!><!UNSUPPORTED!>[]<!><!>)