// !WITH_NEW_INFERENCE
package f

fun <R> h(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>i<!><!>: Int, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>a<!><!>: Any, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>r<!><!>: R, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>f<!><!>: (Boolean) -> Int) = 1
fun <R> h(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>a<!><!>: Any, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>i<!><!>: Int, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>r<!><!>: R, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>f<!><!>: (Boolean) -> Int) = 1

fun test() = <!NI;OVERLOAD_RESOLUTION_AMBIGUITY!><!CANNOT_COMPLETE_RESOLVE!>h<!><!>(1, 1, 1, { <!NI;UNUSED_ANONYMOUS_PARAMETER!><!CANNOT_INFER_PARAMETER_TYPE!><!UNUSED_ANONYMOUS_PARAMETER!>b<!><!><!> -> 42 })
