// !WITH_NEW_INFERENCE
// !DIAGNOSTICS: -CONFLICTING_JVM_DECLARATIONS
package f

fun <R> h(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>f<!><!>: (Boolean) -> R) = 1
fun <R> h(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>f<!><!>: (String) -> R) = 2

fun test() = <!NI;OVERLOAD_RESOLUTION_AMBIGUITY!><!CANNOT_COMPLETE_RESOLVE!>h<!><!>{ <!NI;UNUSED_ANONYMOUS_PARAMETER!><!CANNOT_INFER_PARAMETER_TYPE!><!UNUSED_ANONYMOUS_PARAMETER!>i<!><!><!> -> getAnswer() }

fun getAnswer() = 42

