// !WITH_NEW_INFERENCE
class A() {
    override fun equals(other : Any?) : Boolean = false
}

fun f(): Unit {
    var x: Int? = <!NI;VARIABLE_WITH_REDUNDANT_INITIALIZER!><!VARIABLE_WITH_REDUNDANT_INITIALIZER!>1<!><!>
    x = null
    <!DEBUG_INFO_CONSTANT!>x<!> + 1
    <!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> <!NI;INFIX_MODIFIER_REQUIRED!><!INFIX_MODIFIER_REQUIRED!>plus<!><!> 1
    <!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> <!NI;UNSAFE_OPERATOR_CALL!><!UNSAFE_OPERATOR_CALL!><<!><!> 1
    <!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!><!DEBUG_INFO_CONSTANT!>x<!> += 1<!><!>

    <!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> == 1
    <!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> != 1

    <!NI;EQUALITY_NOT_APPLICABLE!><!EQUALITY_NOT_APPLICABLE!>A() == 1<!><!>

    <!NI;EQUALITY_NOT_APPLICABLE!><!EQUALITY_NOT_APPLICABLE!><!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> === "1"<!><!>
    <!NI;EQUALITY_NOT_APPLICABLE!><!EQUALITY_NOT_APPLICABLE!><!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> !== "1"<!><!>

    <!NI;IMPLICIT_BOXING_IN_IDENTITY_EQUALS!><!IMPLICIT_BOXING_IN_IDENTITY_EQUALS!><!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> === 1<!><!>
    <!NI;IMPLICIT_BOXING_IN_IDENTITY_EQUALS!><!IMPLICIT_BOXING_IN_IDENTITY_EQUALS!><!NI;DEBUG_INFO_CONSTANT!><!DEBUG_INFO_CONSTANT!>x<!><!> !== 1<!><!>

    <!DEBUG_INFO_CONSTANT!>x<!><!NI;UNSAFE_OPERATOR_CALL!><!UNSAFE_OPERATOR_CALL!>..<!><!>2
    <!NI;TYPE_MISMATCH!><!NI;DEBUG_INFO_CONSTANT!><!TYPE_MISMATCH!><!DEBUG_INFO_CONSTANT!>x<!><!><!><!> in 1..2

    val y : Boolean? = true
    <!NI;UNUSED_EXPRESSION!><!UNUSED_EXPRESSION!>false || <!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>y<!><!><!><!>
    <!NI;UNUSED_EXPRESSION!><!UNUSED_EXPRESSION!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>y<!><!> && true<!><!>
    <!NI;UNUSED_EXPRESSION!><!UNUSED_EXPRESSION!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>y<!><!> && <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>1<!><!><!><!>
}