// !WITH_NEW_INFERENCE
fun test() {
  if (<!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>x<!><!> <!NI;COMPARE_TO_TYPE_MISMATCH!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!><!> 0) {

  }
}