// !WITH_NEW_INFERENCE
// NI_EXPECTED_FILE
// See EA-76890 / KT-10843: NPE during analysis
fun lambda(x : Int?) = x?.<!NI;NO_VALUE_FOR_PARAMETER!><!NI;FUNCTION_CALL_EXPECTED!><!NO_VALUE_FOR_PARAMETER!><!TYPE_INFERENCE_NO_INFORMATION_FOR_PARAMETER!><!FUNCTION_CALL_EXPECTED!>let<!><!><!><!><!> <!NI;UNRESOLVED_REFERENCE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>l<!><!> {
    <!CANNOT_INFER_PARAMETER_TYPE!>y<!> ->
    if (<!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>y<!> <!NI;UNREACHABLE_CODE!><!NI;UNRESOLVED_REFERENCE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!> 0<!>) <!NI;UNREACHABLE_CODE!>return<!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>@l<!><!> x<!>
    <!NI;UNREACHABLE_CODE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>y<!><!>
}<!NI;NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!><!NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!>!!<!><!>
