// !WITH_NEW_INFERENCE
operator fun <K, V> MutableMap<K, V>.set(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>k<!><!>: K, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>v<!><!>: V) {}

fun foo(a: MutableMap<String, String>, x: String?) {
    a[x!!] = <!NI;DEBUG_INFO_SMARTCAST!><!DEBUG_INFO_SMARTCAST!>x<!><!>
    a[<!NI;DEBUG_INFO_SMARTCAST!><!DEBUG_INFO_SMARTCAST!>x<!><!>] = x<!NI;UNNECESSARY_NOT_NULL_ASSERTION!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!><!>
}

fun foo1(a: MutableMap<String, String>, x: String?) {
    <!TYPE_INFERENCE_CONFLICTING_SUBSTITUTIONS!>a[<!NI;TYPE_MISMATCH!>x<!>]<!> = x!!
    a[x<!NI;UNNECESSARY_NOT_NULL_ASSERTION!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!><!>] = <!NI;DEBUG_INFO_SMARTCAST!><!DEBUG_INFO_SMARTCAST!>x<!><!>
}