// !WITH_NEW_INFERENCE
import java.util.ArrayList

<!NI;NOT_AN_ANNOTATION_CLASS!>@<!NONE_APPLICABLE!>ArrayList<!><Int>(<!NI;TOO_MANY_ARGUMENTS!>1<!>, <!NI;TOO_MANY_ARGUMENTS!>1<!>)<!> fun b() {}
@<!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>Xoo<!><!>(<!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>x<!><!>) fun c() {}
<!NI;DEPRECATED_JAVA_ANNOTATION!><!DEPRECATED_JAVA_ANNOTATION!>@java.lang.Deprecated(<!NI;UNRESOLVED_REFERENCE!><!NI;TOO_MANY_ARGUMENTS!><!UNRESOLVED_REFERENCE!><!TOO_MANY_ARGUMENTS!>x<!><!><!><!>)<!><!> fun a() {}