// !WITH_NEW_INFERENCE
//KT-1897 When call cannot be resolved to any function, save information about types of arguments

package a

fun bar() {}

fun foo(<!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>i<!><!>: Int, <!NI;UNUSED_PARAMETER!><!UNUSED_PARAMETER!>s<!><!>: String) {}

fun test() {

    bar(<!NI;UNRESOLVED_REFERENCE!><!NI;TOO_MANY_ARGUMENTS!><!UNRESOLVED_REFERENCE!><!TOO_MANY_ARGUMENTS!>xx<!><!><!><!>)

    bar <!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>{ }<!><!>

    foo(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>""<!><!>, <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>1<!><!>, <!NI;UNRESOLVED_REFERENCE!><!NI;TOO_MANY_ARGUMENTS!><!UNRESOLVED_REFERENCE!><!TOO_MANY_ARGUMENTS!>xx<!><!><!><!>)

    foo(<!NI;NAMED_PARAMETER_NOT_FOUND!><!NAMED_PARAMETER_NOT_FOUND!>r<!><!> = <!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>xx<!><!>, i = <!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>""<!><!>, s = "")

    foo(i = 1, <!NI;ARGUMENT_PASSED_TWICE!><!ARGUMENT_PASSED_TWICE!>i<!><!> = 1, s = <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>11<!><!>)

    foo(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>""<!><!>, s = <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>2<!><!>)

    foo(i = <!TYPE_MISMATCH!>""<!>, s = <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>2<!><!>, <!NI;MIXING_NAMED_AND_POSITIONED_ARGUMENTS!><!MIXING_NAMED_AND_POSITIONED_ARGUMENTS!>33<!><!>)

    foo(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>""<!><!>, <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>1<!><!>) <!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>{}<!><!>

    foo(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>""<!><!>, <!NI;CONSTANT_EXPECTED_TYPE_MISMATCH!><!CONSTANT_EXPECTED_TYPE_MISMATCH!>1<!><!>) <!NI;TOO_MANY_ARGUMENTS!><!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>{}<!><!><!> <!MANY_LAMBDA_EXPRESSION_ARGUMENTS!>{}<!>
}