// !WITH_NEW_INFERENCE
fun import() {
    <!NI;FUNCTION_CALL_EXPECTED!><!FUNCTION_CALL_EXPECTED!>import<!><!> <!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>a<!><!><!NI;SYNTAX!><!SYNTAX!>.<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>*<!><!><!NI;SYNTAX!><!><!SYNTAX!><!>
}

fun composite() {
    val <!NI;UNUSED_VARIABLE!><!UNUSED_VARIABLE!>s<!><!> = 13<!NI;OVERLOAD_RESOLUTION_AMBIGUITY!>+<!><!NI;SYNTAX!><!SYNTAX!>~<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>/<!><!>12
}

fun html() {
    <!NI;SYNTAX!><!SYNTAX!><<!><!><!NI;FUNCTION_CALL_EXPECTED!><!FUNCTION_CALL_EXPECTED!>html<!><!><!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>><!><!><!NI;SYNTAX!><!SYNTAX!><<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>/<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!><!NI;COMPARE_TO_TYPE_MISMATCH!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!><!><!NI;SYNTAX!><!><!SYNTAX!><!>
}

fun html1() {
    <!NI;SYNTAX!><!SYNTAX!><<!><!><!NI;FUNCTION_CALL_EXPECTED!><!FUNCTION_CALL_EXPECTED!>html<!><!><!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>><!><!><!NI;SYNTAX!><!SYNTAX!><<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>/<!><!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!><!NI;COMPARE_TO_TYPE_MISMATCH!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!><!><!NI;FUNCTION_CALL_EXPECTED!><!FUNCTION_CALL_EXPECTED!>html<!><!>
}