// !WITH_NEW_INFERENCE
package bar

fun main(args : Array<String>) {
    class Some

    <!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!NI;NO_COMPANION_OBJECT!><!NO_COMPANION_OBJECT!>Some<!><!>[<!NI;SYNTAX!><!><!SYNTAX!><!>]<!><!> <!NI;UNRESOLVED_REFERENCE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>names<!><!> <!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!><!NI;SYNTAX!><!SYNTAX!>=<!><!> ["ads"]<!><!>
}