// !WITH_NEW_INFERENCE
// !DIAGNOSTICS: -UNUSED_EXPRESSION

fun use(a: Any?) = a

fun test() {
    { }<!NI;NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!><!NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!>!!<!><!>
    use({ }<!NI;NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!><!NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION!>!!<!><!>);

    // KT-KT-9070
    <!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>{ }<!><!> <!NI;USELESS_ELVIS!><!USELESS_ELVIS!>?: 1<!><!>
    use(<!NI;TYPE_MISMATCH!>{ 2 }<!> <!NI;USELESS_ELVIS_ON_LAMBDA_EXPRESSION!><!USELESS_ELVIS_ON_LAMBDA_EXPRESSION!>?:<!><!> 1);

    1 <!NI;USELESS_ELVIS!><!USELESS_ELVIS!>?: <!NI;TYPE_MISMATCH!><!NI;UNUSED_LAMBDA_EXPRESSION!><!TYPE_MISMATCH!><!UNUSED_LAMBDA_EXPRESSION!>{ }<!><!><!><!><!><!>
    use(1 <!NI;USELESS_ELVIS!><!USELESS_ELVIS!>?: <!NI;TYPE_MISMATCH!>{ }<!><!><!>)
}